File: musify_backend/app/database/repositories/offline_songs.py
"""
Repositorio para canciones offline
"""
from datetime import datetime
from typing import List, Optional, Dict, Any
from pathlib import Path
import aiosqlite
from app.database.repositories.base import BaseRepository


class OfflineSongsRepository(BaseRepository):
    """Repositorio para gestionar canciones descargadas offline"""
    
    async def get_by_ytid(self, ytid: str) -> Optional[Dict[str, Any]]:
        """
        Obtener canción por ID de YouTube
        
        Args:
            ytid: ID del video de YouTube
            
        Returns:
            Diccionario con datos de la canción o None
        """
        row = await self.fetchone(
            'SELECT * FROM offline_songs WHERE ytid = ?',
            (ytid,)
        )
        
        if row:
            return self._row_to_dict(row)
        return None
    
    async def exists(self, ytid: str) -> bool:
        """
        Verificar si una canción existe en offline
        
        Args:
            ytid: ID del video de YouTube
            
        Returns:
            True si existe, False en caso contrario
        """
        row = await self.fetchone(
            'SELECT 1 FROM offline_songs WHERE ytid = ?',
            (ytid,)
        )
        return row is not None
    
    async def get_all(self) -> List[Dict[str, Any]]:
        """
        Obtener todas las canciones offline
        
        Returns:
            Lista de diccionarios con datos de canciones
        """
        rows = await self.fetchall(
            'SELECT * FROM offline_songs ORDER BY date_added DESC'
        )
        
        return [self._row_to_dict(row) for row in rows]
    
    async def save(
        self,
        ytid: str,
        title: str,
        audio_path: str,
        artist: Optional[str] = None,
        thumbnail: Optional[str] = None,
        artwork_path: Optional[str] = None
    ) -> None:
        """
        Guardar canción offline
        
        Args:
            ytid: ID del video de YouTube
            title: Título de la canción
            audio_path: Ruta al archivo de audio
            artist: Artista (opcional)
            thumbnail: URL del thumbnail (opcional)
            artwork_path: Ruta al archivo de artwork (opcional)
        """
        await self.execute(
            '''INSERT OR REPLACE INTO offline_songs 
               (ytid, title, artist, thumbnail, audio_path, artwork_path, date_added)
               VALUES (?, ?, ?, ?, ?, ?, ?)''',
            (ytid, title, artist, thumbnail, audio_path, artwork_path, 
             datetime.now().isoformat())
        )
        await self.commit()
    
    async def delete(self, ytid: str) -> bool:
        """
        Eliminar canción offline
        
        Args:
            ytid: ID del video de YouTube
            
        Returns:
            True si se eliminó, False si no existía
        """
        await self.execute(
            'DELETE FROM offline_songs WHERE ytid = ?',
            (ytid,)
        )
        await self.commit()
        
        # Verificar si se eliminó algo
        async with await self.execute('SELECT changes()') as cursor:
            result = await cursor.fetchone()
            return result[0] > 0 if result else False
    
    async def count(self) -> int:
        """
        Contar canciones offline
        
        Returns:
            Número de canciones
        """
        row = await self.fetchone('SELECT COUNT(*) FROM offline_songs')
        return row[0] if row else 0
    
    def _row_to_dict(self, row: tuple) -> Dict[str, Any]:
        """Convertir fila de base de datos a diccionario"""
        return {
            'ytid': row[0],
            'title': row[1],
            'artist': row[2],
            'thumbnail': row[3],
            'audioPath': row[4],
            'artworkPath': row[5],
            'dateAdded': row[6],
            'isOffline': True
        }

File: Kordia_backend/src/database/offline-songs.ts
import { BaseRepository } from './base.js';

interface OfflineRow {
  ytid: string; title: string; artist: string | null;
  thumbnail: string | null; audio_path: string;
  artwork_path: string | null; date_added: string;
}

export interface OfflineSongDict {
  ytid: string; title: string; artist: string | null;
  thumbnail: string; audioPath: string;
  artworkPath: string | null; downloaded_at: string;
  isOffline: true;
}

export class OfflineSongsRepository extends BaseRepository {
  getByYtid(ytid: string): OfflineSongDict | undefined {
    const row = this.fetchOne<OfflineRow>(
      'SELECT * FROM offline_songs WHERE ytid = ?', [ytid]
    );
    return row ? this.rowToDict(row) : undefined;
  }

  exists(ytid: string): boolean {
    const row = this.fetchOne<{ '1': number }>('SELECT 1 FROM offline_songs WHERE ytid = ?', [ytid]);
    return !!row;
  }

  getAll(): OfflineSongDict[] {
    return this.fetchAll<OfflineRow>(
      'SELECT * FROM offline_songs ORDER BY date_added DESC'
    ).map(r => this.rowToDict(r));
  }

  save(ytid: string, title: string, audioPath: string, artist?: string | null, thumbnail?: string | null, artworkPath?: string | null): void {
    this.execute(
      `INSERT OR REPLACE INTO offline_songs (ytid, title, artist, thumbnail, audio_path, artwork_path, date_added)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [ytid, title, artist ?? null, thumbnail ?? null, audioPath, artworkPath ?? null, new Date().toISOString()]
    );
  }

  delete(ytid: string): boolean {
    const result = this.execute('DELETE FROM offline_songs WHERE ytid = ?', [ytid]);
    return result.changes > 0;
  }

  count(): number {
    const row = this.fetchOne<{ 'COUNT(*)': number }>('SELECT COUNT(*) FROM offline_songs');
    return row ? row['COUNT(*)'] : 0;
  }

  private rowToDict(r: OfflineRow): OfflineSongDict {
    return {
      ytid: r.ytid,
      title: r.title,
      artist: r.artist,
      thumbnail: r.artwork_path ? `/offline/artwork/${r.ytid}` : r.thumbnail ?? '',
      audioPath: r.audio_path,
      artworkPath: r.artwork_path,
      downloaded_at: r.date_added,
      isOffline: true,
    };
  }
}
